SQLStore: Fix Postgres dialect treating "false" migrator default as true

JacobsonMT opened the pull request:
Previously, when creating a migration you could choose a default value for a new boolean column that looked correct but would be interpreted incorrectly by the Postgres dialect. For example, values such as "false" or "FALSE" would be treated as true by the Postgres dialect.

Example issue: https://github.com/grafana/grafana/commit/531b439cf1d316955b8ab76d0624785de0336d56#diff-da34c94d481aa57b92b9570bb9997a3ec2d40915ccfccaf56684f3afbeb8d68f

This PR refactors how migration dialects determine the Default column value for boolean
type columns. Each dialect now uses the same base code to parse the Default literal and panics if an unknown value is encountered.

So, now AddColumnMigration and AddTableMigration will ensure that across dialects:

- The exact same Default literals will be allowed.
- The literals are converted to equivalent defaults in their DDL.
- An error will be thrown if an invalid literal is provided.

This should cause no change in current migrations as they all use either no default or "0". This was verified by comparing DDL on a fresh instance before and after the change for each of SQLite, MySQL, and Postgres.

JacobsonMT committed:
SQLStore: Fix Postgres dialect treating "false" migrator default as true

Previously, some default migrator values (such as "false" or "FALSE") were
being treated as true by the Postgres dialect.

This refactors how migration dialects determine the Default column value for boolean
type columns. Each dialect now uses the same base code to parse the Default literal
and migrations validate the literals before executing. So, now AddColumnMigration and
AddTableMigration will ensure that across dialects:

- The exact same Default literals will be allowed.
- The literals are converted to equivalent defaults in their DDL.
- An error will be thrown if an invalid literal is provided.

mildwonkey reviewed (approved):
looks reasonable to me

grobinson-grafana commented on pkg/services/sqlstore/migrator/column.go:
```diff
@@ -22,3 +29,33 @@ func (col *Column) String(d Dialect) string {
+// This is meant to ensure that all dialects support the same literals in the same way.
+func ParseBoolStr(s string) (*bool, error) {
+	switch s {
+	case "1", "t", "T", "true", "TRUE", "True":
```
If we want to support all the same inputs as Go, should we replace this switch with `strconv.ParseBool`?

grobinson-grafana commented on pkg/services/sqlstore/migrator/column.go:
```diff
@@ -22,3 +29,33 @@ func (col *Column) String(d Dialect) string {
+		return boolPtr(true), nil
+	case "0", "f", "F", "false", "FALSE", "False":
+		return boolPtr(false), nil
+	case "NULL", "null", "Null", "":
```
Shouldn't this be invalid input? These are not valid bool strings, and a `null` default value should not set `Default`?

JacobsonMT committed:
Fix 'NOT NULL DEFAULT NULL' error for mysql

In Postgres/SQLite, it is valid to have a column with `NOT NULL DEFAULT NULL`,
it's assumed to be equivalent to `NOT NULL`. MySQL though considers this to
be contradictory and throws an error.

Removing support for null literals has some downstream complexity on the
validation solution, so I ended up removing it altogether in favor of a
panic.

JacobsonMT commented:
After some discussion with @grobinson-grafana, we found an issue with MySQL support for `NOT NULL DEFAULT NULL`. In Postgres/SQLite, it is valid to have a column with `NOT NULL DEFAULT NULL`,
it's assumed to be equivalent to `NOT NULL`. MySQL though considers this to
be contradictory and throws an error.

Removing support for null literals has some downstream complexity on the
validation solution, so I ended up removing the validation code altogether in favour of a
panic.

grobinson-grafana commented:
I've also run some tests, here are the results:

```go
mg.AddMigration("test", migrator.NewAddTableMigration(migrator.Table{
	Name: "test",
	Columns: []*migrator.Column{
		{Name: "col1", Type: migrator.DB_Bool, Nullable: false},
		{Name: "col2", Type: migrator.DB_Bool, Nullable: false, Default: "FALSE"},
		{Name: "col3", Type: migrator.DB_Bool, Nullable: false, Default: "f"},
		{Name: "col4", Type: migrator.DB_Bool, Nullable: false, Default: "0"},
		{Name: "col5", Type: migrator.DB_Bool, Nullable: false, Default: "TRUE"},
		{Name: "col6", Type: migrator.DB_Bool, Nullable: false, Default: "t"},
		{Name: "col7", Type: migrator.DB_Bool, Nullable: false, Default: "1"},
		{Name: "col8", Type: migrator.DB_Bool, Nullable: true, Default: "False"},
		{Name: "col9", Type: migrator.DB_Bool, Nullable: true, Default: "f"},
		{Name: "col10", Type: migrator.DB_Bool, Nullable: true, Default: "0"},
		{Name: "col11", Type: migrator.DB_Bool, Nullable: true, Default: "True"},
		{Name: "col13", Type: migrator.DB_Bool, Nullable: true, Default: "t"},
		{Name: "col14", Type: migrator.DB_Bool, Nullable: true, Default: "1"},
		{Name: "col15", Type: migrator.DB_Bool, Nullable: false, Default: "NULL"},
		{Name: "col16", Type: migrator.DB_Bool, Nullable: true, Default: "NULL"},
	},
}))
```

1. NULL is rejected

```
panic: failed to create default value for column 'col15': invalid boolean default value 'NULL' [recovered]
```

2. Defaults work as expected in PostgreSQL

```
postgres=# \d test
                Table "public.test"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 col1   | boolean |           | not null |
 col2   | boolean |           | not null | false
 col3   | boolean |           | not null | false
 col4   | boolean |           | not null | false
 col5   | boolean |           | not null | true
 col6   | boolean |           | not null | true
 col7   | boolean |           | not null | true
 col8   | boolean |           |          | false
 col9   | boolean |           |          | false
 col10  | boolean |           |          | false
 col11  | boolean |           |          | true
 col13  | boolean |           |          | true
 col14  | boolean |           |          | true
```

3. Defaults work as expected in MariaDB (MySQL)

```
MariaDB [grafana]> explain test;
+-------+------------+------+-----+---------+-------+
| Field | Type       | Null | Key | Default | Extra |
+-------+------------+------+-----+---------+-------+
| col1  | tinyint(1) | NO   |     | NULL    |       |
| col2  | tinyint(1) | NO   |     | 0       |       |
| col3  | tinyint(1) | NO   |     | 0       |       |
| col4  | tinyint(1) | NO   |     | 0       |       |
| col5  | tinyint(1) | NO   |     | 1       |       |
| col6  | tinyint(1) | NO   |     | 1       |       |
| col7  | tinyint(1) | NO   |     | 1       |       |
| col8  | tinyint(1) | YES  |     | 0       |       |
| col9  | tinyint(1) | YES  |     | 0       |       |
| col10 | tinyint(1) | YES  |     | 0       |       |
| col11 | tinyint(1) | YES  |     | 1       |       |
| col13 | tinyint(1) | YES  |     | 1       |       |
| col14 | tinyint(1) | YES  |     | 1       |       |
+-------+------------+------+-----+---------+-------+
```

5. Defaults work as expected in sqlite

```
sqlite> PRAGMA table_info([test]);
0|col1|INTEGER|1||0
1|col2|INTEGER|1|0|0
2|col3|INTEGER|1|0|0
3|col4|INTEGER|1|0|0
4|col5|INTEGER|1|1|0
5|col6|INTEGER|1|1|0
6|col7|INTEGER|1|1|0
7|col8|INTEGER|0|0|0
8|col9|INTEGER|0|0|0
9|col10|INTEGER|0|0|0
10|col11|INTEGER|0|1|0
11|col13|INTEGER|0|1|0
12|col14|INTEGER|0|1|0
```

grobinson-grafana commented:
One another test:

```go
mg.AddMigration("another test", migrator.NewAddTableMigration(migrator.Table{
	Name: "another-test",
	Columns: []*migrator.Column{
		{Name: "col1", Type: migrator.DB_Bool, Nullable: false, Default: "Hello, world!"},
	},
}))
```

```
panic: failed to create default value for column 'col1': invalid boolean default value 'Hello, world!' [recovered]
```
